backbone.js

```diff
@@ -658,29 +679,33 @@
     destroy: function(options) {
       options = options ? _.clone(options) : {};
       var model = this;
-      var success = options.success;
       var wait = options.wait;
+      var success = options.success;
+      var error = options.error;
+      delete options.success;
+      delete options.error;
 
       var destroy = function() {
         model.stopListening();
         model.trigger('destroy', model, model.collection, options);
       };
 
-      options.success = function(resp) {
-        if (wait) destroy();
-        if (success) success.call(options.context, model, resp, options);
-        if (!model.isNew()) model.trigger('sync', model, resp, options);
-      };
-
-      var xhr = false;
-      if (this.isNew()) {
-        _.defer(options.success);
-      } else {
-        wrapError(this, options);
-        xhr = this.sync('delete', this, options);
-      }
+      var isNew = model.isNew();
+      var promise = isNew ? Promise.resolve(model.attributes) :
+        wrapError(this.sync('delete', this, options), this, options, error);
       if (!wait) destroy();
-      return xhr;
+      return promise.then(function(resp) {
+        var promise = Promise.resolve(model);
+        if (wait) destroy();
+        if (success)
+          promise = Promise.resolve(success.call(options.context, model, resp, options));
+        if (!isNew)
+          promise = promise.then(function(delivered) {
+            model.trigger('sync', model, resp, options);
```

Comment:
This could be written differently, but the point is that when an existing model is destroyed then the sync event will be triggered after the success callback. Providing a hook to the developer to resolve any rejection by the sync method before the event is triggered.